### PR TITLE
Row level filtering

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: cpp/src/lance/io/reader.cc [hunk 1]
```diff
@@ -261,12 +261,27 @@ ::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> FileReader::ReadBatch(
...
+    int32_t chunk_id,
+    std::shared_ptr<::arrow::UInt64Array> indices) const {
+  return ReadChunk(schema, chunk_id, GetArrayParams(indices));
+}
 
-  return ReadBatch(offset, length, *projection);
+::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> FileReader::ReadChunk(
+    const lance::format::Schema& schema, int32_t chunk_id, const GetArrayParams& params) const {
+  std::vector<std::shared_ptr<::arrow::Array>> arrs;
+  /// TODO: Read field in parallel.
```
Comment (changhiskhan) [inline cpp/src/lance/io/reader.cc:279]: Convert this into a GH issue? feels like an easy win for a new contributor?

File: cpp/src/lance/io/reader.h [hunk 2]
```diff
@@ -73,44 +87,54 @@ class FileReader {
   ::arrow::Result<std::shared_ptr<::arrow::Table>> ReadTable(
       const lance::format::Schema& schema) const;
 
+  /// Consolidate read by offset and read by indices parameters.
+  struct GetArrayParams {
```
Comment (changhiskhan) [inline cpp/src/lance/io/reader.h:91]: "GetArrayParams" sounds like a function that's retrieving some kind of array parameters. This is more like `ArrayTakeParams` or `ArrayReadParams`

File: cpp/src/lance/io/reader.h [hunk 3]
```diff
@@ -73,44 +87,54 @@ class FileReader {
...
+    GetArrayParams(int32_t offset);
```
Comment (changhiskhan) [inline cpp/src/lance/io/reader.h:92]: this would be "read the rest of the array starting from offset"?

Comment (changhiskhan) [inline cpp/src/lance/io/reader.h:92]: If indices are UInt64, should offset/length also be uint64 ?

**Review by changhiskhan — COMMENTED** [review]
looks fine overall. bunch of nits etc. happy to discuss on slack/zoom if it'll be faster.

File: cpp/CMakeLists.txt [hunk 4]
```diff
@@ -116,10 +116,12 @@ if (LANCE_BUILD_BENCHMARKS)
     add_subdirectory(benchmarks)
 endif ()
 
+message("Protobuf llibs: ${Protobuf_LIBRARIES}, ${Protobuf_LITE_LIBRARIES}")
 add_library(lance SHARED ${lance_objects})
 target_link_libraries(lance PUBLIC
+        ${Protobuf_LITE_LIBRARIES}
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/CMakeLists.txt:122]: Good catch.

File: cpp/src/lance/io/project.cc [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  }
+  return std::unique_ptr<Project>(
+      new Project(schema, projected_schema, scan_schema, std::move(filter)));
+}
+
+::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> Project::Execute(
+    std::shared_ptr<FileReader> reader, int32_t chunk_idx) {
+  if (filter_) {
+    auto result = filter_->Execute(reader, chunk_idx);
+    if (!result.ok()) {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/io/project.cc:59]: ARROW_ASSIGN_OR_RAISE does not work well with the auto unbind of `[indices, values]` on line 62.

File: cpp/src/lance/io/project.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+class Project {
+ public:
+  Project() = delete;
+
+  /// Make a Project from the full dataset schema and scan options.
+  static ::arrow::Result<std::unique_ptr<Project>> Make(
+      std::shared_ptr<format::Schema> schema,
+      std::shared_ptr<::arrow::dataset::ScanOptions> scan_options);
+
+  ::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> Execute(std::shared_ptr<FileReader> reader,
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/io/project.h:43]: So in theory, we should have Projection / Filter / LImitOffset all be the subclass of a `PlanNode / ExecNode`. But the signature has not been settled yet IMHO.

File: cpp/src/lance/io/reader.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/io/reader.cc:279]: Done. #43

File: cpp/src/lance/io/scanner.h [hunk 7]
```diff
@@ -18,20 +32,21 @@ class Schema;
 namespace lance::io {
 
 class FileReader;
+class Project;
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/io/scanner.h:35]: ok

[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/io/scanner.h:35]: Usually it is called `Project` tho.

https://arrow.apache.org/docs/cpp/streaming_execution.html
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala#L68

File: cpp/src/lance/io/reader.h (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/io/reader.h:91]: Fixed.

File: cpp/src/lance/io/reader.h (hunk 3 above)
Comment (eddyxu) [reply cpp/src/lance/io/reader.h:92]: to be consistent, lemme change indices to Int32 then.

**Comment by eddyxu** [comment]
The mac build failure is the same as #35 .

File: cpp/src/lance/io/scanner.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (changhiskhan) [reply cpp/src/lance/io/scanner.h:35]: ah ok. if it's the common nomencalture that's fine then

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**